src/generation.rs

```diff
@@ -2154,7 +2149,22 @@ fn key_encoding_field(name: &str, key: &FixedValue) -> EncodingField {
     }
 }
 
-fn encoding_fields(name: &str, ty: SerializingRustType) -> Vec<EncodingField> {
+fn encoding_fields(name: &str, ty: &RustType, include_default: bool) -> Vec<EncodingField> {
+    assert!(CLI_ARGS.preserve_encodings);
+    // TODO: how do we handle defaults for nested things? e.g. inside of a ConceptualRustType::Map
+    let mut encs = encoding_fields_impl(name, ty.into());
+    if include_default && ty.default.is_some() {
+        encs.push(EncodingField {
+            field_name: format!("{}_default_present", name),
+            type_name: "bool".to_owned(),
+            default_expr: "false",
```

Comment:
didn't fully get how it works, why default is false? could you share some context pls?

Reply:
The default behavior, and the idea behind the `.default` notation in CDDL, is to avoid serializing a value if it is equal to the default. When we're keeping track of encodings we keep track of when there is a default value explicitly present and set this to true when that happens to keep the serialization format consistent for round-tripping. This `default_expr` means that unless we happen to encounter the explicitly present default value during deserialization, we should make this encoding value `false` to signify there was no explicitly present default value. This `default_expr` is part of the `EncodingField` and is present for all encoding variables, not just for handling `.default`. e.g. for encoding for `str`s it defaults to `LenEncoding::default()` i.e. canonical encoding (definite length using a minimum of bytes necessary for the length).